aprsmon.8
.TH APRSMON 8 "9 February 1999" Linux "Linux Programmer's Manual"
.SH NAME
aprsmon \- monitor APRS AX.25 traffic for javAPRS
.SH SYNOPSIS
.B aprsmon [-a] [-m] [-r] [-k minutes] [-p port] [-t title] [-i infofile]
.SH DESCRIPTION
.LP
.B Aprsmon
uses SOCK_PACKET facilities to provide a network monitor of all AX.25
UI text traffic heard by the system.  It is based on
listen(1).
.LP
.B Aprsmon
collects (\-m)
or displays standard AX.25 UI text frames in a format similar
to that output by a standard TNC in "Monitor ON" mode and is intended
to be used with programs like javAPRS which wish to see a TNC data
stream over a TCP connection.  When the 
.B aprsmon
master has been running for a while and a new
.B aprsmon
slave starts up, first all saved up reports are sent, then a title
line indicating the beginning of live data is sent, and new reports
are sent as received.
.LP
Compressed position and telemetry reports as generated by the APRS Mic
Encoder are translated into the conventional uncompressed APRS 
Micro-Interface-Module (MIM) telemtry format so
that these stations are visible in javAPRS.  Reports containing a
TheNet X1J4 beacon header are also edited to strip off the header
making these nodes visible in javAPRS as well.
.LP
.SH OPTIONS
.TP 10
.BI \-a
Allow for the monitoring of outgoing frames as well as incoming ones.
.TP 10
.BI \-r
Choose "raw" mode which allows non-printable data through.
.TP 10
.BI "\-p port"
Monitor only those frames received on a particular port, by default all
AX.25 interfaces are monitored.
.TP 10
.BI "-m"
"Master" mode.  Retain monitored data for the number of minutes specified 
with the
.B -k
option in a shared memory segment.
.TP 10
.BI "-k minutes"
Retain monitored packets going back the given number of minutes.  
Default is 30.
.TP 10
.BI "-t title"
Sets the 
.I javAPRS
or
.I aprsd
title string.  Default is "Live data from Linux." For use with aprsd the
format must be something like "aprsmon v2.4.4 padd padb." pada and padb
can be anything, and are ignored but you *must* have 4 words in all.
.TP 10
.BI "-i infofile"
.I Infofile
is the path to the file that the aprsmon master writes the shared memory
segment and semaphore id's into.  Default is "/var/ax25/aprsmon.info."
.SH "javAPRS CONFIGURATION"
Set up one 
.I aprsmon
master running on your system to collect historical data going back
a while.  You will probably want to set up an /etc/rc.d/init.d script for
this.
.sp
.nf
aprsmon \-a \-m&
.fi
.sp
Add an entry into /etc/services with the port number you
want javAPRS to connect on.  For example:
.sp
.nf
aprs		14439/tcp	# APRS monitor on 144.39
.fi
.sp
Add an entry into /etc/inetd.conf for this service:
.sp
.nf
aprs stream tcp nowait root /usr/sbin/aprsmon aprsmon \-a
.fi
.sp
To test, simply "telnet localhost aprs".  To use with javAPRS,
set, for example,
.sp
.nf
<PARAM name = "dataFile1" value = "netc:14439:localhost">
.fi
in your HTML file as described in the 
.B javAPRS
documentation, and then use 
.B appletviewer
to run javAPRS:
.nf
appletviewer mydemo.html
.fi
.sp
.SH "Mic-E to MIM Translation"
The Mic-E reduces the duration of packet noise on a voice frequency
by encoding its data in a compressed form, some of which is binary.
Some APRS-decoding software, notably javAPRS, is unable to deal
with this data.  To solve this problem
.B aprsmon
expands the Mic-E packet into one or two MIM packets as follows:
.sp
.nf
@\fBddhhmm\fPz\fBDDMM.hhN/DDDMM.hhW$cse/spd/E\fP>mon/M\fBm\fP/\fBstatus\fP
T#MIC,\fBaaabbbcccdddeeefffComments
.fi
.sp
.TP 10
.BI ddhhmm
is the UTC day and time the packet was received by the host running 
.B aprsmon.
.TP 10
.BI DDMM.hhN
is the latitude.
.TP 10
.BI /
is the symbol table selection (/ or \\).
.TP 10
.BI DDDMM.hhW
is the longitude.
.TP 10
.BI $
is the symbol.
.TP 10
.BI "cse,spd"
are course and speed.
.TP 10
.BI E
is the type of Mic-Encoder:
.IP
.nf
E - TAPR Mic-Encoder
T - Kenwood TH-D7
.fi
.sp
.TP 10
.BI m
is the status message number, and
.TP 10
.BI status
is the status message number expanded as follows:
.IP
.nf
0 -  Off duty..
1 -  Enroute...
2 -  In Service
3 -  Returning.
4 -  Committed.
5 -  Special...
6 -  PRIORITY..
7 -  EMERGENCY.
.fi
.sp
.TP 10
.BI aaa,bbb,ccc,ddd,eee,fff
are the values of the telemetry channels.
.SH FILES
.BR /etc/ax25/axports
.BR /var/ax25/aprsmon.info
.SH "SEE ALSO"
.BR call (1),
.BR listen (1),
.BR beacon (1),
.BR ax25 (4),
.BR axattach (8),
.BR http://www.bridge.net/~sdimse/javAPRS.html,
.BR http://www.tapr.org,
.BR http://www.aprs.net,
.BR MIC-E.TXT, TELEMTRY.TXT, PROTOCOL.TXT
.SH AUTHOR
.nf
Alan Crosswell, n2ygk@weca.org
.fi


NEWS
Announcing N2YGK's Linux aprsdigi-3.5.1.				September 2012

Release 3.5.1 of N2YGK's advanced APRS digipeater is now available for
free under the terms of the GNU Public License (GPL) at
http://aprsdigi.sourceforge.net.

This release has no new features in the base code from aprsdigi-2.4.5 as the APRS
protocol and aprsdigi code have been stable and operating in production on
a number of digipeaters for many years. However, a number of packaging and documentation
updates have happened:

Added in 3.5.1:

- As usual, the version number has been updated to match the Linux kernel release that 
  Aprsdigi has been built and tested against. (The code actually hasn't had to change since
  kernel 2.4 when ax25 stuff stopped changing.)
- Minor source code changes for netax25 include files no longer required.
- Updated the RPM SPEC file to build aprsdigi for use on Fedora Core 17 or similar systems.
  - Replaced SysV init script (/etc/init.d/aprsdigi) with systemd unit files.
- Added some example configuration files and description of how I build APRS digipeaters
  in ./examples
- Documentation here of 2.4.5 changes that weren't documented at the time (shame on me):

Added in 2.4.5:
- Updated the man page for the "--long" command line options.
- Fixes SIGUSR1 only working once (bug 916083 reported by KE2LJ)
- Updates to the TODO list

Added in 2.4.4:

- Aprsdigi is now a sourceforge.net project.  Please visit aprsdigi.sf.net
  and sign up to be a developer or join the users or developers mailing lists.
- Short (-s) command line options replaced with long ones (--long).  Short
  options are now deprecated and may disappear in a future release.
- Ability to specify a list of ports that get sent a verbatim duplicate copy
  of all packets received on a given port.  Useful for "reflecting" what's
  heard on an RF port to a local LAN for instance.  This is different from
  APRS-IS (e.g. aprsd) since the data is sent with UDP (multicast or unicast).
  Use with caution (or not at all) on RF ports.

Added in 2.4.3:
- UDP/IP (v4 and v6) unicast and multicast (ASM) support.  Useful for 
  connecting aprsdigi's over the Internet, WLANs, etc.
  (Note: this is *not* TCP support as used by APRS-IS, aprsd, etc.)
  Both transparent and "3rd party" tunneling are supported.
- Budlist to permit/deny digipeating of traffic for individuals or groups of
  AX.25 and IP sources.  IP budlists are helpful when using IP multicast!
- Interfaces may be set to transmit- or receive-only.
- Interfaces may be set to disable retransmission of received packets on
  the same interface (one-armed routing).  
- Unix FIFO supported as an interface type.  Mostly useful for debugging.
- Bug fixed that caused aprsdigi to crash upon TRACEn-n callsign insertion
  when the digipeater list was full.

Along with these features:
- Digipeater callsign substitution (e.g. RELAY replaced w/N0CLU-12).
- WIDE, WIDEn-n, TRACE, and TRACEn-n.
- As many radio interfaces as you can plug into your Linux machine
  with flexible gatewaying between them.
- Special handling for Mic-Encoder, Kenwood TH-D7, and TheNet X1J4, including
  SSID-based digipeating.
- Suppression of duplicate and looping packets.
- Works with Linux 2.4 kernel AX.25 network stack.  This means any AX.25
  interface supported by the kernel will work, not just a TNC2 clone.
  Examples include all the BAYCOM interfaces (BAYPAC, PICPAR, PICSER, etc.),
  soundcards, 8530 SCC cards, any serial port KISS TNC, etc.

Aprsdigi is Copyright (c) 1996,1997,1999,2001,2002,2003,2004,2009,2012 Alan Crosswell,
  n2ygk@weca.org


aprsmon.c
/* aprsmon: An incredible simulation of a non-KISS TNC with MONitor ON
 *  (for UI frames only) on a Linux AX.25 interface.  Why?  So it can
 *  be run under inetd and connected to by programs like javAPRS that
 *  want to see a TNC datastream.
 *  (e.g. FROM>TO,DIGI,DIGI...DIGI:text)
 *
 * Runs in two modes:
 *  1. Just spits out received data to stdout.
 *  2. Buffers up some number of minutes worth of received packets into
 *     a segment.  #1 mode dumps contents of this segment before going "live."
 *
 * A socket opened in SOCK_PACKET mode returns frames that are exactly
 * what a KISS TNC sends: a zero, indicating data, followed by the
 * contents of the AX.25 packet.  This program has to decode all that
 * AX.25 junk all over again just to print it out in generic TNC MONitor
 * format.
 *
 * This program is derived from AX.25 utils 2.0.12 listen(8) written by
 * (according to the lsm):
 * 
 *      jsn@cs.nott.ac.uk (Jonathan Naylor)
 *  	tpmannin@cc.hut.fi (Tomi Manninen)
 *      A.Cox@swansea.ac.uk (Alan Cox)
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License, version 2 as
 *  published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * (See the file "COPYING" that is included with this source distribution.)
 * 
 * portions Copyright (c) 1996,1997,1998,1999 Alan Crosswell
 * Alan Crosswell, N2YGK
 * 144 Washburn Road
 * Briarcliff Manor, NY 10510, USA
 * n2ygk@weca.org
 */
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <net/if.h>
#include <netinet/if_ether.h>
#include <netinet/in.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <sys/utsname.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>
#include <signal.h>

#include <linux/ax25.h>

#ifndef HAVE_LIBAX25_EXTENSIONS
#include "libax25ext.h"
#endif
#ifdef USE_SHM
#include "aprsshm.h"
#endif
#include "mic_e.h"

#define	ALEN		6
#define	AXLEN		7
/* SSID mask and various flags that are stuffed into the SSID byte */
#define	SSID		0x1E
#define	HDLCAEB		0x01
#define	REPEATED	0x80

#define	UI		0x03	/* unnumbered information (unproto) */
#define	PID_NO_L3	0xF0	/* no level-3 (text) */

void dump(FILE *f,unsigned char *, int,time_t);
void fmt(const unsigned char *, int, unsigned char **, unsigned char **,time_t);
char *pax25(char *, const unsigned char *);
void cleanup(int);

int master = 0;			/* data collector */
int keepfor = 30;		/* minutes to keep collected data.  */
int raw = 0;			/* print raw data */
char *infofile = "/var/ax25/aprsmon.info";
#ifndef PACKAGE
#define PACKAGE "aprsmon"
#endif

#ifndef VERSION
#define VERSION "$Revision$"
#endif

char *title = "Live data from Linux";

int main(int argc, char **argv)
{
  unsigned char buffer[1500];
  int size;
  int s;
  char *port = NULL, *dev = NULL;
  struct sockaddr sa;
  struct ifreq ifr;
  int proto = ETH_P_AX25;
  time_t now;

#ifdef USE_SHM
#define OPTS "p:amk:i:t:rv"
#else
#define OPTS "p:at:rv"
#endif
  while ((s = getopt(argc, argv, OPTS)) != -1) {
    switch (s) {
#ifdef USE_SHM
    case 'm':
      master++;
      break;
    case 'i':
      infofile = optarg;	/* file name for passing seg/sem info */
      break;
    case 'k':			/* keep for optarg minutes */
      keepfor = atoi(optarg);
      break;
#endif
    case 'p':			/* entry in axports */
      port = optarg;
      break;
    case 'a': /* allows 'hearing' my transmitted packets */
      proto = ETH_P_ALL;
      break;
    case 'r':			/* raw data printing */
      raw++;
      break;
    case 't':
      title = optarg;
      break;
    case 'v':
      printf("aprsmon: %s-%s\n",PACKAGE,VERSION);
      exit(0);
    case '?':
#ifdef USE_SHM
      fprintf(stderr, "Usage: aprsmon [-ar] [-p port] [-m [-k minutes]] [-i infofile] [-t title]\n");
#else
      fprintf(stderr, "Usage: aprsmon [-ar] [-p port] [-t title]\n");
#endif
      return 1;
    }
  }

  if (ax25_config_load_ports() == 0)
    fprintf(stderr, "aprsmon: no AX.25 port data configured\n");
  
  if (port != NULL) {
    if ((dev = ax25_config_get_dev(port)) == NULL) {
      fprintf(stderr, "aprsmon: invalid port name - %s\n", port);
      return 1;
    }
  }
  
  if ((s = socket(AF_PACKET, SOCK_PACKET, htons(proto))) == -1) {
    perror("socket");
    return 1;
  }
  signal(SIGHUP,cleanup);
  signal(SIGINT,cleanup);
  signal(SIGQUIT,cleanup);
  signal(SIGPWR,cleanup);
  signal(SIGPIPE,cleanup);
  
#ifdef USE_SHM
  if (master) {
    if (shm_master(keepfor,infofile))
      return 1;
  } else 
#endif
    {
    struct utsname me;

    uname(&me);
    printf("aprsmon>JAVA:javaMSG  :Linux APRS server (%s-%s by Alan Crosswell, N2YGK) on %s running %s %s\r\n",
	   PACKAGE, VERSION, me.nodename, me.sysname, me.release);
    printf("# %s\r\n",title);
#ifdef USE_SHM
    (void) shm_slave(stdout,infofile);
#endif
    printf("# Live data follows:\r\n");
    printf("aprsmon>JAVA:javaTITLE:%s\r\n",title);
    fflush(stdout);
  }
  for (;;) {
    int asize;

    asize = sizeof(sa);
    if ((size = recvfrom(s, buffer, sizeof(buffer), 0, &sa, &asize)) == -1) {
      perror("recv");
      return 1;
    }
    if (dev != NULL && strcmp(dev, sa.sa_data) != 0)
      continue; 
    
    time(&now);
    if (proto == ETH_P_ALL) {	/* promiscuous mode? */
      strcpy(ifr.ifr_name, sa.sa_data);
      if (ioctl(s, SIOCGIFHWADDR, &ifr) < 0
	  || ifr.ifr_hwaddr.sa_family != AF_AX25)
	continue;		/* not AX25 so ignore this packet */
    }
#ifdef USE_SHM
    if (master)
      collect(buffer,size);
    else
#endif
      dump(stdout,buffer,size,now);
  }
}

void
dump(FILE *f,
     unsigned char *buf,
     int len,
     time_t now)
{
  unsigned char *cp1,*cp2;

  fmt(buf,len,&cp1,&cp2,now);
  if (*cp1) {
    fprintf(f,"%s\r\n",cp1);
  }
  if (*cp2) {
    fprintf(f,"%s\r\n",cp2);
  }
  fflush(f);
}

void
fmt(const unsigned char *buf,
    int len,unsigned char **cp1, 
    unsigned char **cp2,
    time_t now)
{
  static unsigned char buf1[1000],buf2[100];
  char mic1[200],mic2[200];
  int l1,l2;
  char from[10],to[10],digis[100];
  int i,hadlast,l;
  char tmp[15];

  *buf1 = *buf2 = '\0';
  *cp1 = buf1;
  *cp2 = buf2;
  if ((buf[0] & 0xf) != 0)	/* not a kiss data frame */
    return;
  ++buf;
  --len;
  if (len < (AXLEN+ALEN+1))	/* too short */
    return;

  pax25(to,buf);		/* to call */
  pax25(from,&buf[AXLEN]);	/* from call */

  buf+=AXLEN;			/* skip over the from call now... */
  len-=AXLEN;
  *digis = '\0';
  if (!(buf[ALEN] & HDLCAEB)) {	/* is there a digipeater path? */
    for (l=0, buf+=AXLEN, len-=AXLEN, i = 0;
	 i < 6 && len >= 0;
	 i++, len-=AXLEN, buf+=AXLEN) {
      int nextrept = buf[AXLEN+ALEN]&REPEATED;
      if (buf[ALEN]&HDLCAEB)
	nextrept = 0;
      pax25(tmp,buf);
      
      if (*tmp) {
	sprintf(&digis[l],",%s%s",tmp,(buf[ALEN]&REPEATED&&!nextrept)?"*":"");
	++hadlast;
	l += strlen(&digis[l]);
      }
      if (buf[ALEN] & HDLCAEB)
	break;
    }
  }
  buf += AXLEN;
  len -= AXLEN;
  if (len <= 0)
    return;			/* no data after callsigns */
  if (*buf++ == UI && *buf++ == PID_NO_L3) {
    len -= 2;
  } else {
    return;			/* must have UI text to be useful */
  }
  /* now check to see if it's a mic-e frame & rewrite it */
  if (fmt_mic_e(to,buf,len,mic1,&l1,mic2,&l2,now)) {
    /* it was a MIC-E */
    mic1[l1] = mic2[l2] = '\0';
    if (l1)
      sprintf(buf1,"%s>%s%s:%s",from,"APRS",digis,mic1);
    if (l2)
      sprintf(buf2,"%s>%s%s:%s",from,"APRS",digis,mic2);
  } else if (fmt_x1j4(to,buf,len,mic1,&l1,mic2,&l2,now)) {
    /* it was an X1J4 digi */
    mic1[l1] = mic2[l2] = '\0';
    if (l1)
      sprintf(buf1,"%s>%s%s:%s",from,to,digis,mic1);
  } else {
    /* not mic-e/x1j4, so just print the packet without editing */
    sprintf(buf1,"%s>%s%s:",from,to,digis);
    l = strlen(buf1);
    for (i = 0; i < len; i++,l++) {
      buf1[l] = (raw||isprint(buf[i]))?buf[i]:' '; /* keep it clean */
    }
    buf1[l] = '\0';
  }
  return;
}

char *pax25(char *buf, const unsigned char *data)
{
  int i, ssid;
  char *s;
  char c;
  
  s = buf;
  
  for (i = 0; i < ALEN; i++) {
    c = (data[i] >> 1) & 0x7F;
    
    if (!isalnum(c) && c != ' ') {
      strcpy(buf, "[invalid]");
      return buf;
    }
    
    if (c != ' ') *s++ = c;
  }	
  
  if ((ssid = (data[ALEN] & SSID)>>1) != 0)
    sprintf(s, "-%d", ssid);
  else
    *s = '\0';
  
  return(buf);
}

void
cleanup(sig)
int sig;
{
#ifdef USE_SHM
  shm_cleanup(sig);		/* clean up shared memory, etc. */
#endif
  exit(0);
}
